### DISCUSSION TITLE
SKR mini E3 Turbo with DWIN display

### DISCUSSION THREAD

**Discussion by tome9111991** [Jyers/Marlin#837] [General Discussion]
hi
i try to get work the DWIN display with SKR mini E3 Turbo
Any idea how to get working without errors & warring?

in `Marlin\src\pins\lpc1769\pins_BTT_SKR_E3_TURBO.h`

i add

` 
#if ENABLED(DWIN_CREALITY_LCD)

    #ifndef BEEPER_PIN
    #define BEEPER_PIN               EXP1_05_PIN
    #undef SPEAKER
  #endif

    #define BTN_EN1                  EXP1_03_PIN
    #define BTN_EN2                  EXP1_04_PIN
    #define BTN_ENC                  EXP1_06_PIN

    //#define LCD_PINS_RS              EXP1_04_PIN
    //#define LCD_PINS_ENABLE          EXP1_03_PIN
    //#define LCD_PINS_D4              EXP1_05_PIN
  #endif`

The pins from SKR mini looks like
![Screenshot 2021-05-15 111729](https://user-images.githubusercontent.com/57866234/118366168-e5841c00-b59f-11eb-8a05-7902d4fc0534.png)

with Jyers Firmware i got this error
![jyers](https://user-images.githubusercontent.com/57866234/118365360-bcae5780-b59c-11eb-9828-87f45abe1712.png)

With Original Marlin branch i got this warrning i don't know how to fix:

![stock](https://user-images.githubusercontent.com/57866234/118366018-4e1ec900-b59f-11eb-9f2e-9b1b70424420.png)

My config for Jyers branch
[configs.zip](https://github.com/Jyers/Marlin/files/6487623/configs.zip)

**Comment by agalakhov** [comment]
The stock V2 display is connected using UART, not SPI. Look at the pinout written at the LCD board near the connector.
The "Turbo" board is not directly compatible with V2 stock display due to different pinout but it should be possible to reassign pins in software. You have to define LCD_SERIAL to be a software serial connected to the correct pin.

**Reply by tome9111991** [reply]
Dwin has rx and tx pin but im missing the pin on e3 turbo board.
On e3 v2.0 board had tx & rx pin

 if I understand this correctly consider tx and rx to be the software serial
 tx should be EXP1_07_PIN
 rx should be EXP1_08_PIN

**Reply by agalakhov** [reply]
You may convert other pins to RX and TX using software serial by sacrificing some performance. I believe the corresponding pins are at P0.17 and P0.18 but I'm not sure. Or rewire them to the P10 connector, it has RX and TX.

**Reply by agalakhov** [reply]
Yes, you understand this correctly. Just double-check pin assignment.

**Reply by tome9111991** [reply]
maybe i cancel the e3 turbo project i don't know how to use  software serial.

i should buy a mini e3 v2
i can compile without a error or waring 
need only repin the wire

**Comment by tome9111991** [comment]
If someone has free time can try to compile JyersUI for LPC1769.
`default_envs = LPC1769`
`#define MOTHERBOARD BOARD_BTT_SKR_E3_TURBO`

and add in Marlin\src\pins\lpc1769\pins_BTT_SKR_E3_TURBO.h

```
#if ENABLED(DWIN_CREALITY_LCD)

  
    #define BEEPER_PIN               EXP1_10_PIN
    #define BTN_EN1                  EXP1_03_PIN
    #define BTN_EN2                  EXP1_04_PIN
    #define BTN_ENC                  EXP1_06_PIN


 
  #endif
```
with default Marlin Bugfix branch the Ender3 v2 original Display works well with SKR E3 Turbo.
I have still issue to compile with JyersUI got this error:

```
Marlin\src\lcd\dwin\creality_dwin.cpp: In member function 'void Mesh_Settings::Draw_Bed_Mesh(int16_t, uint8_t, uint16_t, uint16_t)':
Marlin\src\lcd\dwin\creality_dwin.cpp:339:27: error: 'MSerialT' {aka 'struct ForwardSerial<MarlinSerial>'} has no member named 'availableForWrite'
  339 |         while (LCD_SERIAL.availableForWrite() < 32) { // wait for serial to be available without blocking and resetting the MCU
      |                           ^~~~~~~~~~~~~~~~~
Marlin\src\lcd\dwin\dwin.cpp: In function 'bool DWIN_Handshake()':
Marlin\src\lcd\dwin\dwin.cpp:98:10: error: no match for 'operator!' (operand type is 'MSerialT' {aka 'ForwardSerial<MarlinSerial>'})
   98 |   while (!LCD_SERIAL && PENDING(millis(), serial_connect_timeout)) { /*nada*/ }
Marlin\src\lcd\dwin\dwin.cpp:98:10: note: candidate: 'operator!(bool)' <built-in>
Marlin\src\lcd\dwin\dwin.cpp:98:10: note:   no known conversion for argument 1 from 'MSerialT' {aka 'ForwardSerial<MarlinSerial>'} to 'bool'
*** [.pio\build\LPC1768\src\src\lcd\dwin\dwin.cpp.o] Error 1
*** [.pio\build\LPC1768\src\src\lcd\dwin\creality_dwin.cpp.o] Error 1
```

**Reply by eohl79** [reply]
I managed to get an SKR E3 Turbo board with stock Creality display working. Besides the LCD pin changes and rewiring, needed the following to fix your compile errors:

in **Marlin/src/inc/Conditionals_LCD.h**
```
#if MB(BTT_SKR_MINI_E3_V1_0, BTT_SKR_MINI_E3_V1_2, BTT_SKR_MINI_E3_V2_0, BTT_SKR_E3_TURBO)
  #define LCD_SERIAL_PORT 1
#else
  #define LCD_SERIAL_PORT 3 // Creality 4.x board
#endif
```

in **Marlin/src/lcd/dwin/creality_dwin.cpp**
`while (!LCD_SERIAL.available()) { // wait for serial to be available without blocking and resetting the MCU `

in **Marlin/src/lcd/dwin/dwin.cpp**
`while (!LCD_SERIAL.connected() && PENDING(millis(), serial_connect_timeout)) { /*nada*/ }`

**Reply by tome9111991** [reply]
@eohl79 
thanks i can compile i got only warrning durring compile.
![Screenshot 2021-07-23 154743](https://user-images.githubusercontent.com/57866234/126791436-096e5d92-6b85-476e-8fde-4af67e82faf8.png)


I will test soon with my display

**EDIT:**
I have tested JyersUI with SKR E3 Turbo it works with Ender3v2 Stock display so far
